### ISSUE TITLE
False positive superfluous-parens

### ISSUE THREAD

**Issue by jolaf** [PyCQA/pylint#4907]
### Bug description

```python
tuple(x for x in ((a, str(a)) for a in ()))
```

### Command used

```shell
pylint test.py
```

### Pylint output

```shell
************* Module test
test.py:1: [C0325(superfluous-parens), ] Unnecessary parens after 'in' keyword
```

### Expected behavior

No errors.

### Pylint version

```shell
pylint 2.10.2
astroid 2.7.2
Python 3.8.10 (default, Jun  2 2021, 10:49:15) 
[GCC 9.4.0]
```

### OS / Environment

Ubuntu 20.04.3 LTS

**Comment by SamyCookie** [comment]
similar issue since 2.10 version with the following example :+1: 
```python
    def __iter__(self):
        return ( (k, getattr(self, k)) for k in self.keys )
 ```

**Comment by DanielNoord** [comment]
@Pierre-Sassoulas This is the result of the following tests and my implementation of solving these in #4784
```python
B = [x for x in ((3, 4))]  # [superfluous-parens]
C = [x for x in ((3, 4) if 1 > 0 else (5, 6))]
D = [x for x in ((3, 4) if 1 > 0 else ((5, 6)))]  # [superfluous-parens]
E = [x for x in ((3, 4) if 1 > 0 else ((((5, 6)))))]  # [superfluous-parens]
```
Looking at these two reports perhaps these are valid cases of double parens and should be removed from the tests, as well as removing from `pylint/checkers/format.py`.
```python
                    if contains_double_parens and tokens[i + 1].string == ")":
                        self.add_message(
                            "superfluous-parens", line=line_num, args=keyword_token
                        )
                        return
                    contains_double_parens = 0
                    continue
```

**Comment by trym-b** [comment]
I think I hit this issue as well, with this code block, altough with the `return` keyword instead of `in`

```python
import numpy as np


def _function_one(var_1: int, var_2: int) -> np.ndarray:
    result = (((var_1 & var_2)) > 0)
    return result.astype(np.float32) # no problem


def _function_two(var_1: int, var_2: int) -> np.ndarray:
    return (((var_1 & var_2)) > 0).astype(np.float32) # Error: Unnecessary parens after 'return' keyword (superfluous-parens)

```

```
test_pylint.py:10:0: C0325: Unnecessary parens after 'return' keyword (superfluous-parens)
```

```
$ pylint --version
pylint 2.10.2
astroid 2.7.3
Python 3.9.5 (default, May 19 2021, 11:32:47) 
[GCC 9.3.0]
```

**Comment by DanielNoord** [comment]
@Pierre-Sassoulas I think you might have missed this. I believe this should be part of 2.11 as it was introduced in 2.10 and is easily fixable (if we do agree that this part of the check should be removed).

**Comment by Pierre-Sassoulas** [comment]
Right, thank you for reminding me. The suggestion in your previous comment makes sense.

**Comment by eli-schwartz** [comment]
I hit this issue too at https://github.com/mesonbuild/meson/blob/dc51740e2cdee171ac6ebc7ed261e5c08730c9a9/mesonbuild/dependencies/mpi.py#L150-L152

```python
            return (f.startswith(('-L', '-l', '-Xlinker')) or
                    f == '-pthread' or
                    (f.startswith('-W') and f != '-Wall' and not f.startswith('-Werror')))
```

The parentheses are needed for grouping, the alternative is a syntax error. This doesn't appear to have been mentioned already in this bug, and it's fixed by the linked ticket, but there doesn't seem to be a test explicitly covering this. Maybe there should be?

**Comment by DanielNoord** [comment]
Yeah, this seems like something different than any of our other tests! I will add an test when I get around to it!
